addons/addon-base-raas/packages/base-raas-services/lib/data-egress/data-egress-service.js

```diff
@@ -124,15 +137,104 @@ class DataEgressService extends Service {
         },
       ],
     };
+    const memberAccountId = await this.getMemberAccountId(requestContext, environment.id);
 
-    const environmentScService = await this.service('environmentScService');
-    const environmentScEntity = await environmentScService.mustFind(requestContext, { id: environment.id });
-    const memberAccount = await environmentScService.getMemberAccount(requestContext, environmentScEntity);
-    await this.addEgressStoreBucketPolicy(requestContext, egressStore, memberAccount.accountId);
+    const lockService = await this.service('lockService');
+    const lockId = `bucket-policy-access-${bucketName}`;
+    await lockService.tryWriteLockAndRun({ id: lockId }, async () => {
+      await this.addEgressStoreBucketPolicy(requestContext, egressStore, memberAccountId);
+    });
 
     return egressStore;
   }
 
+  async terminateEgressStore(requestContext, environmentId) {
+    const enableEgressStore = this.settings.get(settingKeys.enableEgressStore);
+    const by = _.get(requestContext, 'principalIdentifier.uid');
+    if (!enableEgressStore) {
+      throw this.boom.forbidden('Unable to create Egress store since this feature is disabled', true);
+    }
+    const s3Service = await this.service('s3Service');
+    const workspaceId = environmentId;
+    let egressStoreScanResult = [];
+
+    try {
+      egressStoreScanResult = await this._scanner()
+        .limit(1000)
+        .scan()
+        .then(egressStores => {
+          return egressStores.filter(store => store.workspaceId === workspaceId);
+        });
+    } catch (error) {
+      throw this.boom.notFound(`Error in terminating egress store: ${JSON.stringify(error)}`, true);
+    }
+    const egressStoreInfo = egressStoreScanResult[0];
+    const egressStoreStatus = egressStoreInfo.status;
+
+    if (egressStoreStatus.toUpperCase() === PROCESSING_STATUS_CODE) {
+      throw this.boom.forbidden(
+        `Egress store: ${egressStoreInfo.id} is still in processing, workspce can not be terminated before egress request is processed.`,
+        true,
+      );
+    } else if (egressStoreStatus.toUpperCase() === PROCESSED_STATUS_CODE) {
```

Comment:
What is the behavior of the API when the state is neither PROCESSED nor PROCESSING? Do we simply return without any errors. Is that expected?

Reply:
Yes, by design, if the status is in 'CREATED' or 'PENDING', we should do nothing about the egress store and move on terminating the workspace.

Reply:
This is a separate API call for terminating the egress store, right? Shouldn't the API throw an exception for CREATED/PENDING, since technically the call was for terminating the egress store and not for terminating the workspace?

Reply:
Yes, the API should be called for terminate the egress store, however, the message here indicates the workspace termination can not executed because of the egress store status is processing. updated message in commit: 160edb79c93f9d709150bd27c0c46acf198b0662